### PR TITLE
chore: implement DeepCopy for machine configuration

### PR DESCRIPTION
**Pull request by smira**
Resources code extensively uses DeepCopy to prevent in-memory copy of
the resource to be mutated outside of the resource model.

Previous implementation relied on YAML serialization to copy the
machine configuration which was slow, potentially might lead to panics
and it generates pressure on garbage collection.

This implementation uses k8s code generator to generate DeepCopy methods
with some manual helpers when code generator can't handle it.

Signed-off-by: Andrey Smirnov <smirnov.andrey@gmail.com>

### CODE REVIEW

**Comment by smira** [comment]
before this change, copying machine config accounted for 30% of all allocations in `machined` (not in use memory, but just overall allocations)

File: pkg/resources/config/machine_config.go [hunk 1]
```diff
@@ -66,20 +65,10 @@ func (r *MachineConfig) String() string {
...
-	c, err := configloader.NewFromBytes(b)
-	if err != nil {
-		panic(err)
-	}
-
 	return &MachineConfig{
 		md: r.md,
 		spec: &v1alpha1Spec{
-			cfg: c.(*v1alpha1.Config),
+			cfg: r.spec.cfg.(*v1alpha1.Config).DeepCopy(),
```
Comment (smira) [inline pkg/resources/config/machine_config.go:71]: this is the goal of this PR

File: pkg/machinery/config/types/v1alpha1/v1alpha1_types.go [hunk 2]
```diff
@@ -6,16 +6,17 @@
 Package v1alpha1 configuration file contains all the options available for configuring a machine.
 
 To generate a set of basic configuration files, run:
-```bash
-talosctl gen config --version v1alpha1 <cluster name> <cluster endpoint>
-````
+
```
Comment (Unix4ever) [inline pkg/machinery/config/types/v1alpha1/v1alpha1_types.go:9]: This line is actually rendered on our configuration reference page as code block.
Was it something related to deepcopy-gen?

**Review by Unix4ever — DISMISSED** [review]
:cool:

**Review by AlekSi — APPROVED** [review]
LGTM, but it should be rebase now due to my 0.12 folder merge

File: pkg/machinery/config/types/v1alpha1/v1alpha1_types.go (hunk 2 above)
Comment (smira) [reply pkg/machinery/config/types/v1alpha1/v1alpha1_types.go:9]: Ah, I tried to fix the godoc output, I will double check it still renders as code block in the docs.

Comment (smira) [reply pkg/machinery/config/types/v1alpha1/v1alpha1_types.go:9]: yep, it looks good in our docs now and in godoc as well

**Comment by smira** [comment]
> LGTM, but it should be rebased now due to my 0.12 folder merge

updated moving generated docs to 0.12

**Pull request merged**